src/terrain/terrain.js

```diff
@@ -327,6 +328,10 @@ export class Terrain extends Elevation {
         }
     }
 
+    _onSetFeatureStateEvent(event: any) {
+        this._invalidateRenderCache = true;
```

Comment:
There may be a more granular way to invalidate the cache on feature state events. Disabling the entire cache while using _any_ feature state could be a performance hit, as it means we're re-rendering all draping during these events.

I wonder if we could be more granular, at least do that per-tile similar to [_clearRasterFadeFromRenderCache](https://github.com/mapbox/mapbox-gl-js/blob/bcf8cebc594e1df70f1d6de11ca95765418e4020/src/terrain/terrain.js#L870) by identifying which tiles are subject to a feature state change. What do you think?

Reply:
@karimnaaji I like that idea! Do you have thoughts on how to implement it? `_onStyleDataEvent` calls `_clearRenderCacheForTile` when the event passes in a specific source and tile coordinate. Maybe setFeatureState can pass the `sourceId` into the event like the `data` event here:
https://github.com/mapbox/mapbox-gl-js/blob/bcf8cebc594e1df70f1d6de11ca95765418e4020/src/style/style.js#L547

Reply:
How about doing it explicitly from [there](https://github.com/mapbox/mapbox-gl-js/blob/main/src/source/tile.js#L529) instead of with an event?

That's really the deepest code path where we are sure that some data is changing after a feature state.

```js
bucket.update(sourceLayerStates, sourceLayer, availableImages, this.imageAtlas && this.imageAtlas.patternPositions || {});
if (painter.terrain) {
    painter.terrain._clearRenderCacheForTile(sourceCacheId, this.tileID);
}
```
The only thing missing is retrieving the `sourceCacheId`

Reply:
I don't think we can do this without a major overhaul.
`feature-state` works by storing a big global Map(`source_state.js`) that maps `feature-id : feature-state`, and whenever a change occurs it triggers a recalculate paint properties on all features on all tiles.  We'd need change tracking all the way down in `program_configuration`(https://github.com/mapbox/mapbox-gl-js/blob/70df863efb6e55bfdb83c11a3049079b3171eb62/src/data/program_configuration.js#L186) and have to propagate that up.

that would be a cool performance optimization though because if we had that we could update only a small range of the buffer to the gpu.

As a sidenote: Digging for this led me to discover that we can actually optimize performance right here https://github.com/mapbox/mapbox-gl-js/blob/70df863efb6e55bfdb83c11a3049079b3171eb62/src/source/tile.js#L517
We're always `update()` ing on all buckets. But we could query `this.latestFeatureIndex` to figure out if the changed feature-state is  present in the tile and early exit if its not.

Reply:
> How about doing it explicitly from [there](https://github.com/mapbox/mapbox-gl-js/blob/main/src/source/tile.js#L529) instead of with an event?
> 
> That's really the deepest code path where we are sure that some data is changing after a feature state change.
> 
> ```js
> bucket.update(sourceLayerStates, sourceLayer, availableImages, this.imageAtlas && this.imageAtlas.patternPositions || {});
> if (painter.terrain) {
>     painter.terrain._clearRenderCacheForTile(sourceId, this.tileID);
> }
> ```
> 
> The only thing missing is retrieving the `sourceId`

@karimnaaji  I think this gets called for all tiles anyways 😅 . I think the only level of granularity we have right now is invalidating the render cache for an entire source.

Reply:
Thanks @karimnaaji!

Directly calling it as you propose doesn't seem to work, since `painter.terrain` is an instance of `style/terrain.js` and `painter` doesn't have access to the Terrain object we need at `terrain/terrain.js`. What we can do is call an event instead:
```js
bucket.update(sourceLayerStates, sourceLayer, availableImages, this.imageAtlas && this.imageAtlas.patternPositions || {});
if (painter._terrain) {
    painter.style.fire(new Event('setfeaturestate', {coord: this.tileID}));
}
```
As you mention, we're still missing the `source` argument for `terrain._clearRenderCacheForTile(source, coord)`. I'm not sure how we can get that from within a `Tile`. Or is there a different approach here?

Reply:
>@karimnaaji I think this gets called for all tiles anyways 😅 . I think the only level of granularity we have right now is invalidating the render cache for an entire source.

True, but there's already a `dirty` flag that gets [set deeper in this code path](https://github.com/mapbox/mapbox-gl-js/blob/main/src/data/program_configuration.js#L474_L495) if anything has really changed on the feature state buffer, so it should be possible to check that, maybe by being more explicit here and checking this specific status?
```
if (painter._terrain && bucket.programConfigurations.needsUpload) {
  painter.terrain._clearRenderCacheForTile(sourceId, this.tileID);
}
```
>As you mention, we're still missing the source argument for terrain._clearRenderCacheForTile(source, coord). I'm not sure how we can get that from within a Tile

Isn't it `sourceLayerId` defined [here](https://github.com/mapbox/mapbox-gl-js/blob/main/src/source/tile.js#L524)? I might be wrong, but I would check what is already stored in the terrain source cache table to know what is expected as input.

>painter.terrain is an instance of style/terrain.js and painter doesn't have access to the Terrain object we need at terrain/terrain.js.

I don't think so, `painter.terrain` is the instance of the terrain renderer, not the style terrain properties?

Reply:
We could also propagate that dirty flag up to the caller, so that we retrieve it from `bucket.update`, e.g.:

```js
const dirty = bucket.update(...);
if (dirty && terrain) {
  // clear render cache for tile
}
```

Reply:
Thanks @karimnaaji, that seems like a promising approach!

> Isn't it `sourceLayerId` defined [here](https://github.com/mapbox/mapbox-gl-js/blob/main/src/source/tile.js#L524)? I might be wrong, but I would check what is already stored in the terrain source cache table to know what is expected as input.

`sourceLayerId` doesn't match an id, but `painter.style._getSourceCache(bucket.layers[0].source).id` seems to be doing the trick..

Reply:
Here's the code that works:
```js
            if (painter._terrain && bucket.programConfigurations.needsUpload) {
                painter._terrain._clearRenderCacheForTile(painter.style._getSourceCache(bucket.layers[0].source).id, this.tileID);
            }
```
One quirk of this is that `painter._terrain` is not set to a falsey value when terrain is disabled, so this loop continues to run after disabling terrain. That doesn't create any bugs that I know of but I wonder if it might have a performance impact.

Reply:
@SnailBones you should be able to check whether terrain is enabled as well `this._terrain.enabled`, I think that would be what you're looking for?

Reply:
Perfect, thanks!